src/main/java/com/glencoesoftware/omero/ms/image/region/HistogramRequestHandler.java

```diff
@@ -0,0 +1,300 @@
+            // calculated; fall back to plane min/max
+            if (max != 1.0)
+                return new double[] { min, max };
+        }
+        return null;
+    }
+
+    /**
+     * Get the minimum and maximum value of the pixel data
+     *
+     * @param pd
+     *            The {@link PixelData}
+     * @param channel
+     *            The {@link Channel}
+     * @return See above
+     */
+    private double[] getMinMaxFromPixelData(PixelData pd, Channel channel) {
+        double min, max;
+
+        StatsFactory sf = new StatsFactory();
+        double[] pixelMinMax = sf.initPixelsRange(channel.getPixels());
+
+        min = pixelMinMax[1];
+        max = pixelMinMax[0];
+
+        for (int i = 0; i < pd.size(); i++) {
+            min = Math.min(min, pd.getPixelValue(i));
+            max = Math.max(max, pd.getPixelValue(i));
+        }
+
+        return new double[] { min, max };
+    }
+
+    /**
+     * Read through the pixel data and produce histogram data reflecting
+     * the occurrance of pixel values within each bin
+     * @param pd The {@link PixelData} to get the pixel values from
+     * @param channel The {@link Channel} to get the pixel values from
+     * @param minMax The min and max to values to divide into bins
+     * @param imgWidth The sizeX of the pixel data
+     * (not necessarily full resolution)
+     * @param imgHeight The sizeY of the pixel data
+     * (not necessarily full resolution)
+     * @return {@link JsonArray} containing histogram data
+     */
+    public JsonArray getHistogramData(PixelData pd,
+            double[] minMax, int imgWidth, int imgHeight) {
+        int[] counts = new int[histogramCtx.bins];
+
+        double min = minMax[0];
+        double max = minMax[1];
+
+        double range = max - min + 1;
+        double binRange = range / histogramCtx.bins;
+        for (int i = 0; i < pd.size(); i++) {
+            int pdx = i % imgWidth;
+            int pdy = i / imgWidth;
+            if (pdx >= 0 && pdx < (0 + imgWidth) && pdy >= 0 && pdy < (0 + imgHeight)) {
+                int bin = (int) ((pd.getPixelValue(i) - min) / binRange);
+                // if there are more bins than values (binRange < 1) the bin will be offset by -1.
+                // e.g. min=0.0, max=127.0, binCount=256: a pixel with max value 127.0 would go
+                // into bin 254 (expected: 255). Therefore increment by one for these cases.
+                if (bin > 0 && binRange < 1)
+                    bin++;
+
+                if (bin >= 0 && bin < histogramCtx.bins) {
+                    counts[bin]++;
+                } else {
+                    //Pixel Value outside of min/max range
+                    throw new IllegalArgumentException(String.format(
+                            "Image %d has pixel values %.2f outside of [%.2f, %.2f]",
+                            histogramCtx.imageId, pd.getPixelValue(i),
+                            minMax[0], minMax[1]));
```

Comment:
Is it correct the only scenario where this exception could happen would be invalid `StatsInfo` values?

Reply:
Currently I believe so yes.